lib/burnside/flattened.h

```diff
@@ -0,0 +1,264 @@
+  LabelOp(const LabelOp &that);
+  LabelOp &operator=(const LabelOp &that);
+  void setReferenced() const;
+  bool isReferenced() const;
+  LabelRef get() const { return label_; }
+  operator LabelRef() const { return get(); }
+  void dump() const;
+
+private:
+  LabelBuilder &builder_;
+  LabelRef label_;
+};
+
+struct ArtificialJump {};
+constexpr ArtificialJump ARTIFICIAL{};
+
+// a source of an absolute control flow edge
+struct GotoOp
+  : public SumTypeCopyMixin<const parser::CycleStmt *, const parser::ExitStmt *,
+        const parser::GotoStmt *, ArtificialJump> {
+  template<typename A>
+  GotoOp(const A &stmt, LabelRef dest, const Location &source)
+    : SumTypeCopyMixin{&stmt}, target{dest}, source{source} {}
+  explicit GotoOp(LabelRef dest) : SumTypeCopyMixin{ARTIFICIAL}, target{dest} {}
+  void dump() const;
+
+  LabelRef target;
+  Location source;
+};
+
+// control exits the procedure
+struct ReturnOp : public SumTypeCopyMixin<const parser::FailImageStmt *,
+                      const parser::ReturnStmt *, const parser::StopStmt *> {
+  template<typename A>
+  ReturnOp(const A &stmt, const Location &source)
+    : SumTypeCopyMixin{&stmt}, source{source} {}
+  void dump() const;
+
+  Location source;
+};
+
+// two-way branch based on a condition
+struct ConditionalGotoOp
+  : public SumTypeCopyMixin<const parser::Statement<parser::IfThenStmt> *,
+        const parser::Statement<parser::ElseIfStmt> *, const parser::IfStmt *,
+        const parser::Statement<parser::NonLabelDoStmt> *> {
+  template<typename A>
+  ConditionalGotoOp(const A &cond, LabelRef tb, LabelRef fb)
+    : SumTypeCopyMixin{&cond}, trueLabel{tb}, falseLabel{fb} {}
+  void dump() const;
+
+  LabelRef trueLabel;
+  LabelRef falseLabel;
+};
+
+// multi-way branch based on a target-value of a variable
+struct IndirectGotoOp {
+  IndirectGotoOp(
+      const semantics::Symbol *symbol, std::vector<LabelRef> &&labelRefs)
+    : symbol{symbol}, labelRefs{labelRefs} {}
+  void dump() const;
+
+  const semantics::Symbol *symbol;
+  std::vector<LabelRef> labelRefs;
+};
+
+// intrinsic IO operations can return with an implied multi-way branch
+struct SwitchIOOp
+  : public SumTypeCopyMixin<const parser::ReadStmt *, const parser::WriteStmt *,
+        const parser::WaitStmt *, const parser::OpenStmt *,
+        const parser::CloseStmt *, const parser::BackspaceStmt *,
+        const parser::EndfileStmt *, const parser::RewindStmt *,
+        const parser::FlushStmt *, const parser::InquireStmt *> {
+  template<typename A>
+  SwitchIOOp(const A &io, LabelRef next, const Location &source,
+      std::optional<LabelRef> errLab,
+      std::optional<LabelRef> eorLab = std::nullopt,
+      std::optional<LabelRef> endLab = std::nullopt)
+    : SumTypeCopyMixin{&io}, next{next}, source{source}, errLabel{errLab},
+      eorLabel{eorLab}, endLabel{endLab} {}
+  void dump() const;
+
+  LabelRef next;
+  Location source;
+  std::optional<LabelRef> errLabel;
```

Comment:
I would prefer that other `LabelRef`s in these structures be `optional` rather than using a magic value for those that are unspecified.

Reply:
The magic value has a different semantics than "optional". I prefer to keep it distinct rather than to conflate the semantics.